src/compiler/types.ts

```diff
@@ -562,6 +562,10 @@ namespace ts {
     export interface ShorthandPropertyAssignment extends ObjectLiteralElement {
         name: Identifier;
         questionToken?: Node;
+        // used when ObjectLiteralExpression is used in ObjectAssignmentPattern
+        // it is grammar error to appear in actual object initializer
+        equalsToken?: Node;

```

Comment:
I think the idea is to error on the first incorrect token. Also, this controls the span size so it is just 1.